Make text/plugin/media/ua-inline no-quirks documents

mfreed7 opened the pull request:
This change moves these four document types over to be "standards mode" documents.

To implement this, a new Document flag called "parser cannot change the mode" is added to the Document. This flag is checked before the parser tries to change the mode. If the flag is true, no change will be made.

- [x] At least two implementers are interested (and none opposed):
   * Chromium (https://crbug.com/1131185)
   * Mozilla (agreement at [spec triage meeting](https://github.com/whatwg/html/issues/6664#issuecomment-855151234))
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * A test exists for [plain text](https://wpt.fyi/results/html/browsers/browsing-the-web/read-text/load-text-plain.html?label=master&label=experimental&aligned), which I will update.
   * I will try to write some tests for media documents. Plugin might be difficult, as will ua-inline.
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://crbug.com/1131185
   * Firefox: …
   * Safari: …

Closes https://github.com/whatwg/html/issues/5939, Closes https://github.com/whatwg/html/issues/3113


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/6745/browsing-the-web.html" title="Last updated on Jun 23, 2021, 5:26 PM UTC (186b195)">/browsing-the-web.html</a>  ( <a href="https://whatpr.org/html/6745/0e2bdcf...186b195/browsing-the-web.html" title="Last updated on Jun 23, 2021, 5:26 PM UTC (186b195)">diff</a> )
<a href="https://whatpr.org/html/6745/parsing.html" title="Last updated on Jun 23, 2021, 5:26 PM UTC (186b195)">/parsing.html</a>  ( <a href="https://whatpr.org/html/6745/0e2bdcf...186b195/parsing.html" title="Last updated on Jun 23, 2021, 5:26 PM UTC (186b195)">diff</a> )

domenic commented on source:
```diff
@@ -86855,6 +86855,8 @@ new PaymentRequest(&hellip;); // Allowed to use
    object</span> given "<code data-x="">html</code>", <var>type</var>, and
    <var>navigationParams</var>.</p></li>
 
+   <li><p>Set <var>document</var>'s <span data-x="concept-document-mode">mode</span> to "no-quirks".</p></li>
```
```suggestion
   <li><p>Set <var>document</var>'s <span data-x="concept-document-mode">mode</span> to "<code data-x="">no-quirks</code>".</p></li>
```

plus line wrapping, here and elsewhere

domenic reviewed (approved):
LGTM with nits.

I don't know what category `about:blank` falls under (either HTML or ua-inline) but it'd be good to test that as well.

For plugin documents it's worth seeing what people do for same-origin PDFs and whether they actually generate documents with `<embed>`. We might end up deleting that section or converting it into something else as part of the general plugins-are-only-PDFs-these-days deprecation...

mfreed7 commented:
Thanks for the quick review!

The about:blank document isn't included here, as I **would** be afraid of compat issues with that one, given its usage. The spec there is already explicitly quirks-mode: https://html.spec.whatwg.org/#creating-a-new-browsing-context. Step 13 says to mark the document quirks-mode.

I'll see what I can do about pdf/embed tests. You're right that a same-origin PDF should be testable, but that does assume all browsers support PDFs. While true, I don't think it's specified, so writing a WPT feels funny.

I've got the easy ones (text/media) up in a [patch](https://chromium-review.googlesource.com/c/chromium/src/+/2945338) now.

domenic commented:
I did a quick test at https://selective-heliotrope-dumpling.glitch.me/pdf-iframe.html and it looks like for PDFs, Chrome uses the spec's "Page load processing model for content that uses plugins" (with some minor embellishments), while Firefox uses "Page load processing model for inline content that doesn't have a DOM".

Notably, the latter case has an opaque origin and thus cannot be tested. But the former case is same-origin.

So I think a test for this would embed a PDF and pass *either* if accessing `frames[0].document` throws (in which case it's using ua-inline for PDFs) *or* if `frames[0].document.compatMode` is `CSS1Compat`.

Also it'd be ideal to test the difference between `no-quirks` and `limited-quirks` but I'm not sure exactly how to do that....

mfreed7 commented:
Thanks for the quick test! I [made it into a WPT](https://chromium-review.googlesource.com/c/chromium/src/+/2944721) using your suggestion to just pass in the Firefox case. WebKit currently fails this new test.

andreubotella commented on source:
```diff
@@ -86855,6 +86855,9 @@ new PaymentRequest(&hellip;); // Allowed to use
 
+   <li><p>Set <var>document</var>'s <span data-x="concept-document-mode">mode</span> to
+   <code data-x="">no-quirks</code>.</p></li>
+
```
This doesn't work. The default mode for `Document` objects [is already `"no-quirks"`](https://dom.spec.whatwg.org/#ref-for-concept-document-mode%E2%91%A1), and the "create and initialize a `Document` object" algorithm doesn't change that. It's the parser that sets the document's mode to quirks or limited-quirks mode, [depending on the doctype found](https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode:quirks-mode), or if [no doctype is found](https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode:quirks-mode-2).

Therefore, the mode doesn't have to be explicitly set for media, plugin or inline content, because they don't involve a parser; and for text files, the tokenizer must first emit a DOCTYPE token with name `"html"`.

domenic replied:
Eek, thanks for catching this. For text files I guess the parser approach will cause the resulting document have a non-null document.doctype...

andreubotella replied:
I see that Chrome seems to be, at least conceptually, creating documents for text files much like it creates them for media, etc. Changing the algorithm to behave in the same way would probably be cleaner, but I guess that's best as a follow-up.

For now, one way to make the document not be in quirks mode without a DOCTYPE node in the DOM would be to change the document's mode in the "when no more bytes are available..." task.

mfreed7 replied:
Hmm. Ok, let me know where I went off course here:

1. ["Create and initialize a Document object"](https://html.spec.whatwg.org/multipage/browsing-the-web.html#initialise-the-document-object) step 1 says "Let browsingContext be the result of the [obtaining a browsing context to use for a navigation response](https://html.spec.whatwg.org/multipage/origin.html#obtain-browsing-context-navigation)..."
2.  [obtaining a browsing context to use for a navigation response](https://html.spec.whatwg.org/multipage/origin.html#obtain-browsing-context-navigation) step 3 says "Let newBrowsingContext be the result of [creating a new top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-top-level-browsing-context)."
3. [creating a new top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-top-level-browsing-context) step 1 says "Let group be the result of [creating a new browsing context group](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context-group)."
4. [creating a new browsing context group](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context-group) step 3 says "Let browsingContext be the result of [creating a new browsing context](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context) with null, null, and group."
5. [creating a new browsing context](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-browsing-context) step 13 says, and this is key, "Let document be a new Document, **marked as an HTML document in quirks mode**, whose content type..."

mfreed7 replied:
...and I thought it might be better for compat to just directly set the mode to no-quirks, rather than actually adding a `<DOCTYPE>` token to the document. But comments appreciated.

andreubotella replied:
My bad, you're right that the document is initially set to quirks mode. So changing the mode is indeed needed for media/plugin/inline. But for text files, the mode is then set again to quirks mode in the parser.

If adding a `<DOCTYPE>` token is no good, then the mode would have to be switched back to no-quirks at some point after the `<pre>` token is parsed. That's why I suggested changing it back in the task that gets fired when the request is done.

mfreed7 replied:
So the way Blink (and maybe WebKit?) does this is to set the mode in the text/media/etc document constructor, and then ["lock" it](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/text_document.cc;l=36;drc=d27ca1ceb51060e616fb202ead6e6d1b4e02f865). That keeps the `DOCTYPE` detection code in the parser from having any effect, and avoids doing anything at the end of parsing (which could/would cause a re-layout). Perhaps we could add this concept (a "locked" mode) to the spec? It is basically just a flag that you check before allowing changes to the mode.

andreubotella replied:
imo the best way to handle this would be to change the page load processing for text plain to not invoke the parser – it's much cleaner that way.

domenic replied:
The problem with avoiding the parser is that doing so loses the streaming nature of the current parser-based spec. Which is probably observable, given a same-origin iframe loading a giant text file.

The locking mechanism sounds reasonable and is probably the best approach. Waiting until the request is done is probably bad and observable (as it would result in an outer frame observing a change in `frames[0].document.compatMode`).

andreubotella replied:
I don't have any strong objections to the locking mechanism, and it seems like the best option given the circumstances, so +1

mfreed7 replied:
Cool, thanks. I'll try to spec that out then.

mfreed7 replied:
Ok, this PR now has a first cut of this approach. Definitely a bigger change than I was hoping, but not terrible. LMK what you think.

josepharhar commented:
> The spec there is already explicitly quirks-mode: https://html.spec.whatwg.org/#creating-a-new-browsing-context

Oh boy, maybe I should stop using `about:blank` to fiddle with javascript...

domenic reviewed (commented):
I'll note an alternate version here which might not require as many changes to DOM would be:

- Define that each document has a "parser cannot change the mode" boolean, e.g. around https://html.spec.whatwg.org/#overview-of-the-parsing-model or maybe even in https://html.spec.whatwg.org/#the-initial-insertion-mode
- Consult that boolean before performing the mode-setting in https://html.spec.whatwg.org/#the-initial-insertion-mode

mfreed7 commented on source:
```diff
@@ -86931,6 +86940,12 @@ new PaymentRequest(&hellip;); // Allowed to use
+   on <var>document</var>, with <var>new mode</var> set to <code data-x="">no-quirks</code>.</p></li>
+
+   <li><p>Set <var>document</var>'s <span data-x="concept-document-mode-locked">mode locked</span>
+   flag to <code data-x="">true</code>.</p></li>
```
Hmm ok - we currently lock the mode for all of these documents, so I just implemented the spec the same way. Any harm in keeping it locked for all document types?

domenic replied:
Yeah, I think it's pretty confusing for readers to add something to the spec that in actuality has no impact.

mfreed7 commented:
> I'll note an alternate version here which might not require as many changes to DOM would be:
> 
> * Define that each document has a "parser cannot change the mode" boolean, e.g. around https://html.spec.whatwg.org/#overview-of-the-parsing-model or maybe even in https://html.spec.whatwg.org/#the-initial-insertion-mode
> * Consult that boolean before performing the mode-setting in https://html.spec.whatwg.org/#the-initial-insertion-mode

Hmm, so you'd add this in the HTML spec, not in DOM? I saw this comment after I ended up [filing the DOM PR](https://github.com/whatwg/dom/pull/989). The biggest changes are here in HTML, and the DOM change is pretty minimal. So let me know what you think as-is.

annevk commented:
I hadn't read this discussion when I read the DOM PR. From this discussion it seems that over time HTML might not use the HTML parser for these type of documents in which case this problem would not come up, right? From that perspective a solution local to HTML seems preferable, as it can be more easily refactored over time.

mfreed7 commented:
> I'll note an alternate version here which might not require as many changes to DOM would be:
> 
> * Define that each document has a "parser cannot change the mode" boolean, e.g. around https://html.spec.whatwg.org/#overview-of-the-parsing-model or maybe even in https://html.spec.whatwg.org/#the-initial-insertion-mode
> * Consult that boolean before performing the mode-setting in https://html.spec.whatwg.org/#the-initial-insertion-mode

Alright, I updated the PR to use this suggestion instead of making changes to DOM.  I'm not sure I added the flag to Document in the right spot, it feels kind of tossed in. Let me know if there's a better spot for it. I also updated the non-normative descriptions for parse errors to be correct, but I'm not sure we need to do that? And finally, I removed the "mode locks" from all other document types except text mode.

@annevk and @domenic, PTAL, thanks!

mfreed7 commented:
Hmm, upon a re-read, I think I misunderstood this comment:

> This is not necessary for the non-text cases, I don't think...

I believe we still do need to change the mode, you were just saying we didn't need to "lock" it, since the parser doesn't try to change it. If that's correct, I can add back setting the mode to no-quirks for plugin/media/ua-inline documents.

domenic commented on source:
```diff
@@ -86855,6 +86855,11 @@ new PaymentRequest(&hellip;); // Allowed to use
+   <li><p>Set the <var>document</var>'s <span>parser cannot change the mode flag</span> to true.</p></li>
+
+   <li><p>Set <var>document</var>'s <span data-x="concept-document-mode">mode</span>
+   flag to <code data-x="">no-quirks</code>.</p></li>
```
```suggestion
   to "<code data-x="">no-quirks</code>".</p></li>
```

mfreed7 replied:
All done.

domenic commented on source:
```diff
@@ -104076,16 +104081,16 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
-     placed as a document preamble, the parser sets the <code>Document</code> to <span>quirks
-     mode</span>.</p>
+     placed as a document preamble, and if the <span>parser cannot change the mode flag</span> is false,
+     the parser sets the <code>Document</code> to <span>quirks mode</span>.</p>
```
Yeah, hmm, I think there's no need to explain this in the syntax errors table, since DOCTYPEs won't show up in text documents where this would occur.

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -109103,10 +109112,13 @@ document.body.appendChild(text);
+  <h6>The <span>parser cannot change the mode flag</span></h6>
 
-  <h5 id="parsing-main-inhtml">The rules for parsing tokens in HTML content</h5>
+  <p>A <code>Document</code> object has an associated <dfn for="Document">parser cannot change the mode flag</dfn> (a boolean). It is initially false.</p>
```
Maybe throw this under the initial insertion mode, instead of giving it its own section?

mfreed7 replied:
Ok, done.

domenic reviewed (commented):
> I believe we still do need to change the mode, you were just saying we didn't need to "lock" it, since the parser doesn't try to change it. If that's correct, I can add back setting the mode to no-quirks for plugin/media/ua-inline documents.

Yeah, I think per your reasoning at https://github.com/whatwg/html/pull/6745#discussion_r646926769, we indeed still need to change the mode.

mfreed7 commented:
> Yeah, I think per your reasoning at [#6745 (comment)](https://github.com/whatwg/html/pull/6745#discussion_r646926769), we indeed still need to change the mode.

Ok, I put those three back. And address the rest of your comments. Let me know if I missed anything, thanks!

domenic reviewed (approved):
LGTM!

Can you update the OP with a link to the tests that have been updated for text and media? It's fine to not do plugin and ua-inline.

Similarly go ahead and file bugs on WebKit and Gecko.

@annevk want to give things a final look? If not I'll merge tomorrow; everything seems nice to me.
